compressors/lz77.py

```diff
@@ -0,0 +1,705 @@
+                is chosen. Defaults to 64.
+            window_initialization (List, optional): initialize window (this is like side information
+                or dictionary in zstd parlance). The same initialization should be used for the decoder.
+
+        """
+        self.min_match_length = min_match_length
+        self.max_num_matches_considered = max_num_matches_considered
+        self.window = []
+        self.dict = (
+            {}
+        )  # map from substr of length min_match_length to list of positions where it occurs
+        self.window_indexed_till = 0  # pointer telling up to what point the window has been indexed
+        # if window_indexed_till = 100, that means all substrings starting at 0,1,2,...,100-min_match_length+1
+        # have been indexed
+
+        # if window_initialization is provided, update window and index it
+        if window_initialization is not None:
+            self.window = list(window_initialization)
+            self.index_window_upto_pos(len(self.window))
+
+    def reset(self):
+        # reset the window and the index
+        self.window = []
+        self.dict = {}
+        self.window_indexed_till = 0
+
+    def insert_substring_into_dict(self, substr: Tuple, start_pos: int):
+        """Insert substring into the dict (mapping substring to positions where
+        it occurs in the window).
+
+        Args:
+            substr (Tuple): tuple of length min_match_length
+            start_pos (int): position of substr in window
+        """
+        if substr in self.dict:
+            self.dict[substr].append(start_pos)
+        else:
+            self.dict[substr] = [start_pos]
+
+    def index_window_upto_pos(self, end_pos: int):
+        """Index all tuples of min_match_length in self.window[:end_pos]
+        into the dict. The last tuple to be indexed will start at end_pos-min_match_length+1
+        This uses self.window_indexed_till to ensure we do not reindex things already indexed.
+
+        Args:
+            end_pos (int): end position in window to index
+        """
+        for end_pos_substr in range(self.window_indexed_till, end_pos + 1):
+            start_pos_substr = end_pos_substr - self.min_match_length
+            if start_pos_substr < 0:
+                continue
+            substr = tuple(self.window[start_pos_substr:end_pos_substr])
+            self.insert_substring_into_dict(substr, start_pos_substr)
+        self.window_indexed_till = end_pos + 1
+
+    def encode_lz77_sequences(self, lz77_sequences: LZ77Sequence):
+        """Perform entropy encoding of the LZ77 sequences and return the encoded bitarray.
+
+        We simply use the Elias Delta code for encoding all three, but first we preprocess a bit to
+        optimize the encoding.
+
+        Args:
+            encoded_bitarray (BitArray): encoded bit array
+        """
+        literal_counts = [l.literal_count for l in lz77_sequences]
```

Comment:
surprised that just Elias-Gamma coding for all the params of the `lz77_sequences` turns out to be not-so-bad for encoding.

Reply:
I was inspired by the deflate standard which uses a fixed encoding.